File: app/app/src/main/java/com/example/hakonsreader/views/Tag.java
package com.example.hakonsreader.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.core.content.ContextCompat;

import com.example.hakonsreader.R;
import com.example.hakonsreader.databinding.TagBinding;
import com.squareup.picasso.Picasso;

public class Tag extends LinearLayout {
    private static final String TAG = "Tag";
    
    private TagBinding binding;
    private String text;
    private int textColor;
    private int fillColor;

    private boolean textColorOverriden = false;

    public Tag(Context context) {
        this(context, null, 0);
    }
    public Tag(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }
    public Tag(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.Tag, 0, 0);
        try {
            text = a.getString(R.styleable.Tag_text);
            textColor = a.getColor(R.styleable.Tag_textColor, ContextCompat.getColor(context, R.color.text_color));
            fillColor = a.getColor(R.styleable.Tag_textColor, ContextCompat.getColor(context, R.color.background));
        } finally {
            a.recycle();
        }

        binding = TagBinding.inflate(LayoutInflater.from(context), this, true);

        // With elevation the card will have a weird looking shadow (only visible on light mode)
        binding.cardView.setCardElevation(0);
    }

    /**
     * Adds a view to the flair
     *
     * <p>If the tag has multiple views ensure they are added in the correct order</p>
     *
     * <p>Padding is added to the view</p>
     */
    public void add(View view) {
        view.setPadding(5, 0, 5, 1);
        binding.tags.addView(view);
    }

    /**
     * Sets the fill/background color of the tag
     *
     * @param color The color resource ID
     */
    public void setFillColor(int color) {
        binding.cardView.setCardBackgroundColor(color);
    }

    /**
     * Sets the fill/background color of the tag
     *
     * @param hexColor A hex string representing a color. Note: If this is set to "transparent"
     *                 then the text color will be overridden and always be set to the current themes
     *                 text color
     */
    public void setFillColor(String hexColor) {
        if (hexColor.equals("transparent")) {
            // Getting the transparent value out of a color is apparently harder (if there is no transparent
            // it still returns 255? so set it here only, not in setFillColor(int))
            setFillColor(ContextCompat.getColor(getContext(), R.color.background_with_alpha));
            textColor = ContextCompat.getColor(getContext(), R.color.text_color);
            textColorOverriden = true;
            // The elevation shadow will still be visible on light backgrounds
            binding.cardView.setCardElevation(0f);
        } else {
            setFillColor(Color.parseColor(hexColor));
        }
    }

    /**
     * Sets the text color for the tag. If {@link Tag#setFillColor(String)} is set with "transparent"
     * then this will override the text color with the themes text color
     *
     * @param textColor The color to set for the text
     */
    public void setTextColor(int textColor) {
        if (!textColorOverriden) {
            this.textColor = textColor;
        }
    }

    /**
     * Adds text to the tag
     *
     * <p>Note the text color must be set with {@link Tag#setTextColor(int)} before this is called</p>
     *
     * @param text The text to add. If this is empty nothing is added to the tag
     */
    public void addText(String text) {
        if (text.isEmpty()) {
            return;
        }

        // If the last item in the tags is a TextView, add the text to that view instead
        // of creating a new view
        int tagsAdded = binding.tags.getChildCount();
        if (tagsAdded > 0) {
            View last = binding.tags.getChildAt(tagsAdded - 1);

            if (last instanceof TextView) {
                TextView asTextView = (TextView) last;
                asTextView.append(" " + text);

                return;
            }
        }

        TextView tv = new TextView(getContext());

        tv.setText(text);
        tv.setTextColor(textColor);
        tv.setTextSize(TypedValue.COMPLEX_UNIT_PX, getContext().getResources().getDimension(R.dimen.tagTextSize));

        tv.setEllipsize(TextUtils.TruncateAt.END);
        tv.setMaxLines(1);

        this.add(tv);
    }

    /**
     * Adds an image to the tag
     *
     * @param imageURL The URL to the image
     */
    public void addImage(String imageURL) {
        int size = (int)getContext().getResources().getDimension(R.dimen.tagIconSize);
        ImageView iv = new ImageView(getContext());
        Picasso.get()
                .load(imageURL)
                .resize(size, size)
                .into(iv);

        this.add(iv);
    }
}

File: app/app/src/main/java/com/example/hakonsreader/views/Tag.kt
package com.example.hakonsreader.views

import android.content.Context
import android.graphics.Color
import android.text.TextUtils
import android.util.AttributeSet
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.example.hakonsreader.R
import com.example.hakonsreader.databinding.TagBinding
import com.squareup.picasso.Picasso

class Tag @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr) {
    private val binding: TagBinding = TagBinding.inflate(LayoutInflater.from(context), this, true)
    private var textColor = 0
    private var textColorOverridden = false

    init {
        val a = context.theme.obtainStyledAttributes(attrs, R.styleable.Tag, 0, 0)
        try {
            val defaultTextColor = "#000000"
            textColor = a.getColor(R.styleable.Tag_textColor, Color.parseColor(defaultTextColor))
            val defaultFillColor = "#EAEAEA"
            val fillColor = a.getColor(R.styleable.Tag_fillColor, Color.parseColor(defaultFillColor))
            setFillColor(fillColor)
            val text = a.getString(R.styleable.Tag_text)
            if (text != null) {
                addText(text)
            } else {
                if (isInEditMode) {
                    addText("FaZe Clan fan")
                }
            }
        } finally {
            a.recycle()
        }

        // With elevation the card will have a weird looking shadow (only visible on light mode)
        binding.cardView.cardElevation = 0f
    }

    /**
     * Clears all views from the tag
     */
    fun clear() {
        binding.tags.removeAllViews()
    }

    /**
     * Adds a view to the flair
     *
     *
     * If the tag has multiple views ensure they are added in the correct order
     *
     *
     * Padding is added to the view
     */
    fun add(view: View) {
        view.setPadding(5, 0, 5, 1)
        binding.tags.addView(view)
    }

    /**
     * Sets the fill/background color of the tag
     *
     * @param color The color resource ID
     */
    fun setFillColor(color: Int) {
        binding.cardView.setCardBackgroundColor(color)
    }

    /**
     * Sets the fill/background color of the tag
     *
     * @param hexColor A hex string representing a color. Note: If this is set to "transparent"
     * then the text color will be overridden and always be set to the current themes
     * text color
     */
    fun setFillColor(hexColor: String) {
        if (hexColor == "transparent") {
            // Getting the transparent value out of a color is apparently harder (if there is no transparent
            // it still returns 255? so set it here only, not in setFillColor(int))
            setFillColor(ContextCompat.getColor(context, R.color.background_with_alpha))
            textColor = ContextCompat.getColor(context, R.color.text_color)
            textColorOverridden = true
            // The elevation shadow will still be visible on light backgrounds
            binding.cardView.cardElevation = 0f
        } else {
            setFillColor(Color.parseColor(hexColor))
        }
    }

    /**
     * Sets the text color for the tag. If [Tag.setFillColor] is set with "transparent"
     * then this will override the text color with the themes text color
     *
     * @param textColor The color to set for the text
     */
    fun setTextColor(textColor: Int) {
        if (!textColorOverridden) {
            this.textColor = textColor
        }
    }

    /**
     * Adds text to the tag
     *
     * Note the text color must be set with [Tag.setTextColor] before this is called
     *
     * @param text The text to add. If this is empty nothing is added to the tag
     */
    fun addText(text: String) {
        if (text.isEmpty()) {
            return
        }

        // If the last item in the tags is a TextView, add the text to that view instead
        // of creating a new view
        val tagsAdded = binding.tags.childCount
        if (tagsAdded > 0) {
            val last = binding.tags.getChildAt(tagsAdded - 1)
            if (last is TextView) {
                last.append(" $text")
                return
            }
        }

        TextView(context).run {
            this.text = text
            setTextColor(textColor)
            setTextSize(TypedValue.COMPLEX_UNIT_PX, context.resources.getDimension(R.dimen.tagTextSize))
            ellipsize = TextUtils.TruncateAt.END
            maxLines = 1
            add(this)
        }
    }

    /**
     * Adds an image to the tag
     *
     * @param imageURL The URL to the image
     */
    fun addImage(imageURL: String?) {
        val size = context.resources.getDimension(R.dimen.tagIconSize).toInt()

        ImageView(context).run {
            Picasso.get()
                    .load(imageURL)
                    .resize(size, size)
                    .into(this)
            add(this)
        }
    }
}